### ISSUE TITLE
Active academic research: request for particular idea :muscle: 

### ISSUE THREAD

**Issue by attilamester** [radareorg/radare2#22571]
## DFS on merged agC and agR to obtain instruction list

* we are analyzing BODMAS samples
* for a sample, we create a call graph by merging both agC and agR
  * there are some cases when nodes appear in one, and not the other
  * we want to analyze all possible addresses
* we want to implement a simple depth-first-search traversal starting from the entry, and gather a list of instructions, with regard to function call 
  * meaning that if 
  node1 [mov, push, call, xor, adc, ret] calls 
  node2 [mov, mov, ret] 
  then the DFS order will be 
  **[ ... mov, push, mov, mov, xor, adc ...]**
* are there by chance any r2 commands already for this?

### Problem 1: cannot correlate merge(agC, agR) with pdf

* the `call` set from pdf is not the same as the call set from merge(agC, agR)
* consider fcn.1 calls fcn.2 by creating a thread
* pdf command output for fcn.1:
```
push fcn.2
call sym.imp.KERNEL32.dll_CreateThread
```
* agC command:
```
fcn.1 -> sym.imp.KERNEL32.dll_CreateThread
```
* agR command:
```
fcn.1 -> fcn.2
fcn.1 -> sym.imp.KERNEL32.dll_CreateThread
```

* extract from our code:
```
    def get_node_calls_from_instructions(self) -> List[str]:
        """
        Correlate the node's calls given the `agCd` and `agRd` with the node's instructions given by `pdfj`
        :return: labels of the nodes called by the instructions of the given node, in the order of its instructions
        """
        node_calls_labels = [n.label for n in self.get_calls()] # based on merge on agCd and agRd
        node_calls_labels_from_instructions = []

        for i in self.instructions:
            if "call" in i.mnemonic or "jmp" in i.mnemonic:
                ind = i.disasm.index(i.mnemonic)
                callee = i.disasm[ind + len(i.mnemonic):]
                for label in node_calls_labels:
                    _label = label.replace("unk.", "")
                    if _label in callee:
                        node_calls_labels_from_instructions.append(label)
                        break

        # if set(node_calls_labels_from_instructions) != set(node_calls_labels):
        #     raise Exception(
        #         f"For {self} could not find some callees: {node_calls_labels} != {node_calls_labels_from_instructions}")
        return node_calls_labels_from_instructions
```

### Problem 2: emulator would be needed? eg. mov eax, fcn.1; [...] call eax

* our code should track the registers, memory addresses, mov/push operations in order to be able to tell if we hit a `call eax` instruction, which entry it means from agC/agR
* is it really needed to re-implement the wheel? an emulator? are we missing something? :D 

### Example

* real-world case from an executable
* entry node:
```
[0x00401000]> pdf
            ;-- section..code:
            ;-- eip:
┌ 157: entry0 ();
│           0x00401000      682c000000     push 0x2c                   ; ',' ; 44 ; [00] -r-x section size 20480 named .code
│           0x00401005      6800000000     push 0                      ; int c
│           0x0040100a      6850024300     push 0x430250               ; void *s
│           0x0040100f      e8fc4f0000     call sub.MSVCRT.dll_memset  ; void *memset(void *s, int c, size_t n)
│           0x00401014      83c40c         add esp, 0xc
│           0x00401017      6800000000     push 0                      ; int argc
│           0x0040101c      e8f54f0000     call main                   ; int main(int argc, char **argv, char **envp)
│           0x00401021      a354024300     mov dword [0x430254], eax   ; [0x430254:4]=0
│           0x00401026      6800000000     push 0
│           0x0040102b      6800100000     push 0x1000
│           0x00401030      6800000000     push 0                      ; DWORD flOptions
│           0x00401035      e8e24f0000     call sub.KERNEL32.dll_HeapCreate ; HANDLE HeapCreate(DWORD flOptions, SIZE_T dwInitialSize, SIZE_T dwMaximumSize)
│           0x0040103a      a350024300     mov dword [0x430250], eax   ; [0x430250:4]=0
│           0x0040103f      e86c540000     call fcn.004064b0
│           0x00401044      e8af500000     call fcn.004060f8
│           0x00401049      6840004300     push str.NtReleaseMutant    ; 0x430040 ; "NtReleaseMutant"
│           0x0040104e      6850004300     push str.ntdll              ; 0x430050 ; "ntdll"
│           0x00401053      68ffffffff     push 0xffffffff             ; -1 ; uint32_t arg_ch
│           0x00401058      e8c2500000     call fcn.0040611f
│           0x0040105d      50             push eax                    ; int32_t arg_8h
│           0x0040105e      e8f9500000     call fcn.0040615c
│           0x00401063      a378024300     mov dword [0x430278], eax   ; [0x430278:4]=0
│           0x00401068      681e004300     push str.snxhk.dll          ; 0x43001e ; "snxhk.dll"
│           0x0040106d      68ffffffff     push 0xffffffff             ; -1 ; uint32_t arg_ch
│           0x00401072      e8a8500000     call fcn.0040611f
│           0x00401077      a36c024300     mov dword [0x43026c], eax   ; [0x43026c:4]=0
│           0x0040107c      bb050000c0     mov ebx, 0xc0000005
│           0x00401081      f7db           neg ebx
│           0x00401083      81c39f104000   add ebx, 0x40109f
│           0x00401089      6a7b           push dword [0x430278]
│           0x0040108b      6a56           pop ecx
│           0x0040108d      ff1578024300   call ecx
│           0x00401093      031d6c024300   add ebx, dword [0x43026c]   ; [0x43026c:4]=0
│           0x00401099      01c3           add ebx, eax
└           0x0040109b      ffe3           jmp ebx
[0x00401000]> 
```
* extract from the agCd:
![image](https://github.com/radareorg/radare2/assets/38431467/47fa2965-9be8-4e12-b2c4-79a380e99531)
* how could we know which call instruction from the `pdf` links to which function from the `agCd` image?

## Note

* maybe we are missing the forest from the trees :face_with_head_bandage: 
* maybe we focused too much time on these agC, agR, pdf commands
* maybe there are other tools as well
* any hints, ideas are welcome :muscle:

**Comment by attilamester** [comment]
## Update

### Problem 3: pdf and agC/agR

* how can we detect by looking at pdf output of function `0x0004cdad` that this function has a call to `fcn.00007d80` ?
* fcn.00007d80 seems to be present only in agR and in pdf's comment section after a call

```
[0x00000000]> s 0x0004cdad
[0x0004cdad]> pdf
            ; CALL XREF from fcn.0004d0fd @ 0x4d421(x)
            ; CALL XREFS from fcn.0004debd @ 0x4decd(x), 0x4df4c(x)
            ; CALL XREF from fcn.0004e32d @ +0x88(x)
┌ 149: fcn.0004cdad (int64_t arg4, int64_t arg_8h);
│           ; arg int64_t arg4 @ rcx
│           ; arg int64_t arg_8h @ rsp+0x70
│           ; var int64_t var_30h @ rsp+0x30
│           ; var int64_t var_38h @ rsp+0x38
│           ; var int64_t var_3ch @ rsp+0x3c
│           ; var int64_t var_40h @ rsp+0x40
│           ; var signed int64_t var_44h @ rsp+0x44
│           ; var signed int64_t var_48h @ rsp+0x48
│           ; var int64_t var_54h @ rsp+0x54
│           0x0004cdad      894c2408       mov dword [arg_8h], ecx     ; arg4
│           0x0004cdb1      4883ec68       sub rsp, 0x68
│           0x0004cdb5      c7442438f35b.  mov dword [var_38h], 0x5bf3 ; [0x5bf3:4]=0xfc77287a
│           0x0004cdbd      488d442438     lea rax, [var_38h]
│           0x0004cdc2      4889442430     mov qword [var_30h], rax
│           0x0004cdc7      8b442438       mov eax, dword [var_38h]
│           0x0004cdcb      ffc0           inc eax
│           0x0004cdcd      89442454       mov dword [var_54h], eax
│           0x0004cdd1      488b4c2430     mov rcx, qword [var_30h]
│           0x0004cdd6      8b01           mov eax, dword [rcx]
│           0x0004cdd8      99             cdq
│           0x0004cdd9      8b4c2454       mov ecx, dword [var_54h]
│           0x0004cddd      f7f9           idiv ecx
│           0x0004cddf      89442440       mov dword [var_40h], eax
│           0x0004cde3      ff15c47d0000   call qword [fcn.00054bad]   ; [0x54bad:8]=0x4ad82da2c4da64ab
│           0x0004cde9      488d44243c     lea rax, [var_3ch]
│           0x0004cdee      f0ff08         lock dec dword [rax]
│           0x0004cdf1      c74424440000.  mov dword [var_44h], 0
│       ┌─< 0x0004cdf9      eb0a           jmp 0x4ce05
│       │   ; CODE XREF from fcn.0004cdad @ 0x4ce35(x)
│      ┌──> 0x0004cdfb      8b442444       mov eax, dword [var_44h]
│      ╎│   0x0004cdff      ffc0           inc eax
│      ╎│   0x0004ce01      89442444       mov dword [var_44h], eax
│      ╎│   ; CODE XREF from fcn.0004cdad @ 0x4cdf9(x)
│      ╎└─> 0x0004ce05      837c244405     cmp dword [var_44h], 5
│      ╎┌─< 0x0004ce0a      7d2b           jge 0x4ce37
│      ╎│   0x0004ce0c      c74424480000.  mov dword [var_48h], 0
│     ┌───< 0x0004ce14      eb0a           jmp 0x4ce20
│     │╎│   ; CODE XREF from fcn.0004cdad @ 0x4ce33(x)
│    ┌────> 0x0004ce16      8b442448       mov eax, dword [var_48h]
│    ╎│╎│   0x0004ce1a      ffc0           inc eax
│    ╎│╎│   0x0004ce1c      89442448       mov dword [var_48h], eax
│    ╎│╎│   ; CODE XREF from fcn.0004cdad @ 0x4ce14(x)
│    ╎└───> 0x0004ce20      837c244804     cmp dword [var_48h], 4
│    ╎┌───< 0x0004ce25      7d0e           jge 0x4ce35
│    ╎│╎│   0x0004ce27      ff15907d0000   call qword [0x00054bbd]     ; [0x54bbd:8]=0x4adc2da2c4dc64ab
│    ╎│╎│   0x0004ce2d      ff15827d0000   call qword [fcn.00054bb5]   ; [0x54bb5:8]=0x4ada2da2c4d264ab
│    └────< 0x0004ce33      ebe1           jmp 0x4ce16
│     │││   ; CODE XREF from fcn.0004cdad @ 0x4ce25(x)
│     └└──< 0x0004ce35      ebc4           jmp 0x4cdfb
│       │   ; CODE XREF from fcn.0004cdad @ 0x4ce0a(x)
│       └─> 0x0004ce37      ff15807d0000   call qword [0x00054bbd]     ; [0x7d80:8]=0xa00d8b100499b0a1 ; fcn.00007d80
│           0x0004ce3d      4883c468       add rsp, 0x68
└           0x0004ce41      c3             ret
[0x0004cdad]> agCd ~0004cdad
  "0x0004cdad" [label="fcn.0004cdad" URL="fcn.0004cdad/0x0004cdad"];
  "0x0004cdad" -> "0x00054bad" [color="#61afef" URL="fcn.00054bad/0x00054bad"];
  "0x0004cdad" -> "0x00054bbd" [color="#61afef" URL="unk.0x54bbd/0x00054bbd"];
  "0x0004cdad" -> "0x00054bb5" [color="#61afef" URL="fcn.00054bb5/0x00054bb5"];
  "0x0004debd" -> "0x0004cdad" [color="#61afef" URL="fcn.0004cdad/0x0004cdad"];
  "0x0004cdad" [label="fcn.0004cdad" URL="fcn.0004cdad/0x0004cdad"];
  "0x0004d0fd" -> "0x0004cdad" [color="#61afef" URL="fcn.0004cdad/0x0004cdad"];
  "0x0004cdad" [label="fcn.0004cdad" URL="fcn.0004cdad/0x0004cdad"];
[0x0004cdad]> agRd ~0004cdad
"fcn.0004cdad" [URL="fcn.0004cdad", color="lightgray", label="fcn.0004cdad"]
"fcn.0004cdad" -> "0x00005bf3"
"fcn.0004cdad" -> "0x00007dc4"
"fcn.0004cdad" -> "fcn.00054bad"
"fcn.0004cdad" -> "0x0004ce05"
"fcn.0004cdad" -> "0x0004ce37"
"fcn.0004cdad" -> "0x0004ce20"
"fcn.0004cdad" -> "0x0004ce35"
"fcn.0004cdad" -> "0x00007d90"
"fcn.0004cdad" -> "0x00054bbd"
"fcn.0004cdad" -> "0x00007d82"
"fcn.0004cdad" -> "fcn.00054bb5"
"fcn.0004cdad" -> "0x0004ce16"
"fcn.0004cdad" -> "0x0004cdfb"
"fcn.0004cdad" -> "fcn.00007d80"
"fcn.0004cdad" -> "0x00054bbd"
"fcn.0004d0fd" -> "fcn.0004cdad"
"fcn.0004debd" -> "fcn.0004cdad"
"fcn.0004debd" -> "fcn.0004cdad"
[0x0004cdad]>
```

**Comment by attilamester** [comment]
TLDR; is there a method to get the same output as `pdfj`, but when there is a call, to have a RVA (so not dword [..], not label, but pure RVA), also include in te output?

**Comment by attilamester** [comment]
Closing this issue since its too vague, and a more precise description of the problem is tracked in https://github.com/radareorg/radare2/issues/22576
